pydoctor/templatewriter/util.py

```diff
@@ -19,13 +19,22 @@ def templatefile(filename):
 def templatefilepath(filename):
     return FilePath(templatefile(filename))
 
-def taglink(o: model.Documentable, label: Optional[str] = None) -> Tag:
+def taglink(o: model.Documentable, page_url: str, label: Optional[str] = None) -> Tag:
     if not o.isVisible:
         o.system.msg("html", "don't link to %s"%o.fullName())
+
     if label is None:
         label = o.fullName()
+
+    url = o.url
+    if url.startswith(page_url + '#'):
+        # When linking to an item on the same page, omit the path.
+        # Besides shortening the HTML, this also avoids the page being reloaded
+        # if the query string is non-empty.
+        url = url[len(page_url):]
```

Comment:
Do we have test for this ?

I am checking the demo at https://pydoctor--351.org.readthedocs.build/en/351/docformat/epytext/demo/epytext_demo.demo_epytext_module.html#demo_fields_docstring_arguments

In the source I see

```html
<p>This function can be used in conjuction with
<a href="epytext_demo.demo_epytext_module.html#demo_typing_arguments"><code>demo_typing_arguments</code>
</a> to find an arbitrary function's zeros.
</p>
```
I was expected HREF to be just `#demo_typing_arguments`

and there is hte type `conjuction` vs conju**n**ction

Reply:
We have test coverage ;) ...but that coverage is meaningless since the produced links aren't checked.

I did verify the output on two projects, so I'm wondering what is different about the demo that it doesn't work there. To be continued...

Reply:
> and there is hte type `conjuction` vs conju**n**ction

Maybe it would be a useful feature to have a spell check option for docstrings?

It's not as simple as it sounds though: to avoid false positives, we'd have to exclude for example class/function/variable names, but how those are embedded in docstrings differs per documentation format. I think we need a DOM that is doc format independent, not only for spell check but also for other features like the automatic construction of the summary line.

Possible options for the DOM would include:
- docutils
- HTML subset with well-defined semantics
- something custom

A problem with using an HTML subset is that while all our documentation format modules can convert to HTML, getting them to produce the same nodes in the same situations would probably be a significant effort. For example, if one doc format generates `<code>` and another generates `<span class="code">`, that would make it hard to process the DOM.

A problem with using something other than HTML is that we might run into problems if we ever want to support Markdown as a docstring format, since Markdown allows embedding raw HTML. On the other hand, even if we do pick HTML as the DOM, the raw HTML inside Markdown is not guaranteed to stick to our subset, so we'd need to either normalize the HTML or preserve embedded HTML as-is in a node type that we ignore during processing.

Designing something custom is always an option we can fall back on, but it shouldn't be the first thing we try, in my opinion (avoid re-inventing the wheel). Even if we do end up with a custom solution, knowing what didn't work in the previously tried approaches would help create a better design.

Reply:
> I did verify the output on two projects, so I'm wondering what is different about the demo that it doesn't work there. To be continued...

The reason turns out to be simple: the generation of that particular link is not done via `util.taglink()` (which I updated), but using a direct `tags.a` construction instead (which was left unchanged).

However, the solution is not as simple:

- should `resolve_identifier_xref()` call `taglink()`? this would require changing the `DocstringLinker` interface, since it would have to return a tag rather than a URL
- or should the "omit the path when link target is on the same page" behavior be refactored into a separate function?

Those two options lead to different output: `taglink()` also forces the `code` CSS class and adds a `data-type` attribute.

Reply:
The `data-type` attribute was added in #90 and is there to tell `doc2dash` whether the linked documentable is a class, method, variable etc. The only file that the pydoctor parser of `doc2dash` searches for these `data-type` attributes is `nameIndex.html`, so we could only add them for that file and reduce the output size a bit.

I also opened a ticket to discuss whether we can export the index information in a more structured way than `data-` attributes in HTML, for example using the Intersphinx inventory: hynek/doc2dash#99.

Reply:
Using both a `code` CSS class and `<code>` tags was adding complexity without any benefits, so I'll be moving everything over to use only `<code>` tags.

Reply:
For spell checking and in general, yes. It would be nice to have an internal DOM and then just have convertors from Epytext or RST into this pydoctor DOM... but I am not sure if it's worth the trouble right now.


-----------

I have never used doc2dash and in fact I am not using any other API docs tools. Just reading the docs from local repo or in web browser.
But I think that making pydoctor easier for Dash would be nice.

Reply:
I think it's worth a try to convert epytext to `docutils.node`: if that works, we don't need to design our own DOM. And we might get better Sphinx integration as well.

Dash itself seems to be Mac-only, but I installed Zeal under Linux and it looks like a useful application. The documentation itself is the same as in a browser, but loading and searching is very fast.